eclair-core/src/test/scala/fr/acinq/eclair/integration/basic/payment/BlindedPaymentSpec.scala

```diff
@@ -0,0 +1,319 @@
+      .modify(_.features.activated).using(_ + (RouteBlinding -> Optional))
+      .modify(_.channelConf.channelFlags.announceChannel).setTo(!testData.tags.contains(PrivateChannels))
+    val bobParams = nodeParamsFor("bob", ByteVector32(hex"7620226fec887b0b2ebe76492e5a3fd3eb0e47cd3773263f6a81b59a704dc492"))
+      .modify(_.features.activated).using(_ + (RouteBlinding -> Optional))
+      .modify(_.features.activated).usingIf(testData.tags.contains(RouteBlindingDisabledBob))(_ - RouteBlinding)
+      .modify(_.channelConf.channelFlags.announceChannel).setTo(!testData.tags.contains(PrivateChannels))
+    val carolParams = nodeParamsFor("carol", ByteVector32(hex"ebd5a5d3abfb3ef73731eb3418d918f247445183180522674666db98a66411cc"))
+      .modify(_.features.activated).using(_ + (RouteBlinding -> Optional))
+      .modify(_.features.activated).using(_ + (KeySend -> Optional))
+      .modify(_.features.activated).usingIf(testData.tags.contains(RouteBlindingDisabledCarol))(_ - RouteBlinding)
+      .modify(_.channelConf.channelFlags.announceChannel).setTo(!testData.tags.contains(PrivateChannels))
+
+    val f = ThreeNodesFixture(aliceParams, bobParams, carolParams, testData.name)
+    createChannels(f)
+    f
+  }
+
+  override def cleanupFixture(fixture: FixtureParam): Unit = {
+    fixture.cleanup()
+  }
+
+  private def createChannels(f: FixtureParam): Unit = {
+    import f._
+
+    alice.watcher.setAutoPilot(watcherAutopilot(knownFundingTxs(alice, bob, carol)))
+    bob.watcher.setAutoPilot(watcherAutopilot(knownFundingTxs(alice, bob, carol)))
+    carol.watcher.setAutoPilot(watcherAutopilot(knownFundingTxs(alice, bob, carol)))
+
+    connect(alice, bob)
+    connect(bob, carol)
+
+    val channelId_ab = openChannel(alice, bob, 500_000 sat).channelId
+    val channelId_bc_1 = openChannel(bob, carol, 100_000 sat).channelId
+    val channelId_bc_2 = openChannel(bob, carol, 100_000 sat).channelId
+
+    eventually {
+      assert(getChannelData(alice, channelId_ab).asInstanceOf[DATA_NORMAL].shortIds.real.isInstanceOf[RealScidStatus.Final])
+      assert(getChannelData(bob, channelId_bc_1).asInstanceOf[DATA_NORMAL].shortIds.real.isInstanceOf[RealScidStatus.Final])
+      assert(getChannelData(bob, channelId_bc_2).asInstanceOf[DATA_NORMAL].shortIds.real.isInstanceOf[RealScidStatus.Final])
+    }
+  }
+
+  def createInvoice(recipient: MinimalNodeFixture, amount: MilliSatoshi, routes: Seq[ReceivingRoute], sender: TestProbe): Bolt12Invoice = {
+    val offer = Offer(None, "test", recipient.nodeParams.nodeId, Features.empty, recipient.nodeParams.chainHash)
+    val invoiceReq = InvoiceRequest(offer, amount, 1, Features.empty, randomKey(), recipient.nodeParams.chainHash)
+    sender.send(recipient.paymentHandler, MultiPartHandler.ReceiveOfferPayment(randomKey(), offer, invoiceReq, routes, recipient.router))
+    val invoice = sender.expectMsgType[Bolt12Invoice]
+    assert(invoice.nodeId != recipient.nodeParams.nodeId)
```

Comment:
Why this check? You create an offer with `recipient.nodeParams.nodeId`, it would actually be a bug if the corresponding invoice didn't have the same node id.

Reply:
Yes, the key provided must be the private corresponding to the node id in the offer (if not, the signature will be invalid and the payer will cancel the payment). I don't think we would gain anything with `OfferDetails(nodeKey: PrivateKey, offer: Offer)`, we just need to make it clearer that the key must match the offer, and maybe add a `require` in the `Bolt12Invoice` constructor.

Reply:
Thanks for the details, that makes sense. This isn't related to the current PR though, is it? I think it makes sense to make that change in the PR where we will actually generate offers and invoices in response to invoice requests, doesn't it? Because that's where we'll start generating these random receiver keys and will have to store them in the DB.

Reply:
Fine